### PR TITLE
Dynamo driver beginning

### PR DESCRIPTION
**Pull request by niels-qup**
Not finished, just for impression sake.

### CODE REVIEW

File: database/dynamo/transaction_writer_test.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+import (
+	"fmt"
+	"testing"
+)
+
+func TestTransactionWriter_Delete(t *testing.T) {
+	connection := TestConnection{}
+	dynamoClient := CreateNewQupDynamo(connection)
+
+	transaction, err := dynamoClient.Transaction("mockTable", ExampleRecord{})
```
Comment (pcittadini) [inline database/dynamo/transaction_writer_test.go:12]: Interesting how my vet is throwing a warning here:
'transaction' may have 'nil' or other unexpected value as its corresponding error variable may be not 'nil'

But I don't see how this can happen in the definition of Transaction

Comment (niels-qup) [reply database/dynamo/transaction_writer_test.go:12]: It's because it returns a pointer and i guess the vet is too stupid to realise that this pointer is always set

File: database/dynamo/reflect.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+	"github.com/queueup-dev/qup-io/reflection"
+	"reflect"
+	"regexp"
+	"strings"
+)
+
+type DynamoTableDefinition struct {
+	PrimaryKey          string
+	Fields              []string
+	GlobalSearchIndices map[string][]string
```
Comment (pcittadini) [inline database/dynamo/reflect.go:14]: Shouldn't this be GlobalSecondaryIndex?

File: database/dynamo/client.go [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+}
+
+type Connection interface {
+	GetItem(input *dynamodb.GetItemInput) (*dynamodb.GetItemOutput, error)
+	DeleteItem(input *dynamodb.DeleteItemInput) (*dynamodb.DeleteItemOutput, error)
+	Scan(input *dynamodb.ScanInput) (*dynamodb.ScanOutput, error)
+	TransactWriteItems(input *dynamodb.TransactWriteItemsInput) (*dynamodb.TransactWriteItemsOutput, error)
+	Query(input *dynamodb.QueryInput) (*dynamodb.QueryOutput, error)
+}
+
```
Comment (daniel-qup) [inline database/dynamo/client.go:20]: It might be nice to add the following:
var _ Connection = &dynamodb.DynamoDB{}

In order to:
 - communicate the intention of the interface to other developers
 - as a check on whether or not the DynbamoDB struct implements said interface, this might change in the future.

File: database/dynamo/transaction_writer_test.go (hunk 1 above)
Comment (daniel-qup) [reply database/dynamo/transaction_writer_test.go:12]: For performance I don't think it is needed to return a pointer when dealing with small structs. And for bigger structs, most of the time the parts that are big are placed behind a pointer.

It looks like the reason you return a pointer is because an error can be returned in the Transaction method. Or do you need a pointer for testing purposes? Other than that, it's perfectly fine to return an empty struct. (As for example relfect.ValueOf() does.)

However I'm still unsure on what my preference here is.

File: database/dynamo/reflect.go (hunk 2 above)
Comment (niels-qup) [reply database/dynamo/reflect.go:14]: Could be local as well, but i get your feedback. Maybe we should just make it `Indices` as i don't treat local and global differently.

File: database/dynamo/client.go (hunk 3 above)
Comment (niels-qup) [reply database/dynamo/client.go:20]: Makes sense.

File: database/dynamo/reflect.go (hunk 2 above)
Comment (niels-qup) [reply database/dynamo/reflect.go]: √

File: database/dynamo/client.go (hunk 3 above)
Comment (niels-qup) [reply database/dynamo/client.go:22]: √

**Review by pcittadini — APPROVED** [review]
I cannot add any other remark. I think minor adjustments may come as we use it.
